File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.py
import sys
sys.setrecursionlimit(10**6)


class Node:
    def __init__(self, value, num):
        self.left = None
        self.right = None
        self.value = value
        self.num = num


def insert(parent_node, new_node):
    if not parent_node:
        return new_node

    if new_node.value < parent_node.value:
        parent_node.left = insert(parent_node.left, new_node)
    else:
        parent_node.right = insert(parent_node.right, new_node)

    return parent_node


def preorder(root, ls):
    if not root:
        return
    ls.append(root.num)
    preorder(root.left, ls)
    preorder(root.right, ls)


def postorder(root, ls):
    if not root:
        return
    postorder(root.left, ls)
    postorder(root.right, ls)
    ls.append(root.num)


def solution(nodeinfo):
    nodeinfo = [(a, *b) for a, b in enumerate(nodeinfo)]

    nodeinfo.sort(key=lambda x: (-x[2], x[1]), reverse=True)

    root = None
    while nodeinfo:
        num, value, *_ = nodeinfo.pop()
        new_node = Node(value, num + 1)
        root = insert(root, new_node)

    result = [[], []]
    preorder(root, result[0])
    postorder(root, result[1])
    return result


# print(solution([[5, 3], [11, 5], [13, 3], [3, 5], [6, 1], [1, 3], [8, 6], [7, 2], [2, 2]]))

File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.java
import java.util.*;
import java.util.stream.Collector;
import java.util.stream.Collectors;

class Node{
    public int value;
    public int num;
    public Node left;
    public Node right;

    public Node(int value, int num) {
        this.value = value;
        this.num = num;
    }

    @Override
    public String toString() {
        return "Node{" +
                "value=" + value +
                ", num=" + num +
                '}';
    }
}



class Solution {
    public Node insert(Node parentNode, Node newNode){
        if (parentNode == null) {
            return newNode;
        }

        if(newNode.value < parentNode.value){
            parentNode.left = insert(parentNode.left, newNode);
        }else{
            parentNode.right = insert(parentNode.right, newNode);
        }

        return parentNode;
    }

    public List<Integer> preList = new ArrayList<>();
    public List<Integer> postList = new ArrayList<>();

    public void preOrder(Node root){
        if (root == null) {
            return;
        }
        preList.add(root.num);
        preOrder(root.left);
        preOrder(root.right);
    }

    public void postOrder(Node root) {
        if (root == null) {
            return;
        }
        postOrder(root.left);
        postOrder(root.right);
        postList.add(root.num);
    }

    public int[][] solution(int[][] nodeinfo) {
        for (int i = 0; i < nodeinfo.length; i++) {
            nodeinfo[i] = new int[]{i + 1, nodeinfo[i][0], nodeinfo[i][1]};
        }

        LinkedList<int[]> list = Arrays.stream(nodeinfo)
                .sorted(Comparator.comparing((int[] s) -> -s[2]).thenComparing(s -> s[1]))
                .collect(Collectors.toCollection(LinkedList::new));


        Node root = null;
        while (!list.isEmpty()) {
            int[] ints = list.pollFirst();
            root = insert(root, new Node(ints[1], ints[0]));
        }

        preOrder(root);
        postOrder(root);

        return new int[][]{preList.stream().mapToInt(i -> i).toArray(), postList.stream().mapToInt(i -> i).toArray()};
    }
}